Add documentation of Hubble Exporter (#27610)
cilium-team/cilium: Documentation/observability/hubble-exporter.rst

- ``hubble-export-allowlist``: specify an allowlist as JSON encoded FlowFilters to Hubble exporter.

- ``hubble-export-denylist``: specify a denylist as JSON encoded FlowFilters to Hubble exporter.

- ``hubble-export-fieldmask``: specify a list of fields to use for field masking in Hubble exporter.

Filters
-------

You can use ``hubble`` CLI to generated required filters (see `Specifying Raw
Flow Filters`_ for more examples).

.. _Specifying Raw Flow Filters: https://github.com/cilium/hubble#specifying-raw-flow-filters

For example, to filter flows with verdict ``DENIED`` or ``ERROR``, run:

.. code-block:: shell-session

    $ hubble observe --verdict DROPPED --verdict ERROR --print-raw-filters
    allowlist:
    - '{"verdict":["DROPPED","ERROR"]}'

Then paste the output to ``hubble-export-allowlist`` in ``cilium-config``
Config Map:

.. code-block:: shell-session

    kubectl -n kube-system patch cm cilium-config --patch-file=/dev/stdin <<-EOF
    data:
      hubble-export-allowlist: '{"verdict":["DROPPED","ERROR"]}'
    EOF

You can do the same to selectively filter data. For example, to filter all flows in the
``kube-system`` namespace, run:

.. code-block:: shell-session

>     $ hubble observe --not --namespace kube-system --print-raw-filters

Reviewer: Nice! TIL about `--not` option

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

***************************
Configuring Hubble exporter
***************************

**Hubble Exporter** is an option of ``cilium-agent`` that allows to write
Hubble flows to a file. This can be later consumed as logs. It supports file

> rotation, size limits, filters and field masks.

Reviewer 2: Use clear, plain language, comma nit, specific is better than general
```suggestion
**Hubble Exporter** is a feature of ``cilium-agent`` that lets you write
Hubble flows to a file for later consumption as logs. Hubble Exporter supports file
rotation, size limits, filters, and field masks.
```

Basic Configuration
===================

**Hubble Exporter** is configured via Config Map (default name:
``cilium-config``). This feature is disabled until a value of

> ``hubble-export-file-path`` is set.

Reviewer 2: Use clear plain language, use active voice
```suggestion
Configure **Hubble Exporter** with Config Map. The feature has a default name of
``cilium-config``. Hubble Exporter is disabled until you set a file path value for
``hubble-export-file-path``.
```

.. code-block:: shell-session

    kubectl -n kube-system patch cm cilium-config --patch-file=/dev/stdin <<-EOF
    data:
      hubble-export-file-path: "/var/run/cilium/hubble/events.log"
    EOF

Restart ``cilium-agent`` to apply the change:

.. code-block:: shell-session

    kubectl -n kube-system rollout restart ds/cilium-agent

> Verify that change was applied (it can take a few minutes before first flow is

Reviewer 2: Grammar nit. I don't love the passive voice here, but I think it's the best choice.
```suggestion
Verify that the change was applied (it can take a few minutes before the first flow is
```

logged):

.. code-block:: shell-session

    kubectl -n kube-system exec ds/cilium -- tail -f /var/run/cilium/hubble/events.log

> Configure your logging solution to consume logs from the specified location.

Reviewer 2: Specific is better than general
```suggestion
Configure your logging solution to consume logs from your Hubble export file path.
```

> Other configuration options:

Reviewer 2: Clarity
```suggestion
Other configuration options include:
```

- ``hubble-export-file-max-size-mb``:
  Size in MB at which to rotate Hubble export file. (default 10)

- ``hubble-export-file-max-backups``:
  Number of rotated Hubble export files to keep. (default 5)

- ``hubble-export-file-compress``:

>   Compress rotated Hubble export files.

Reviewer 2: I'd rather see these pairs on single lines.

Is `hubble-export-file-compress` a Boolean? What values does it accept?
```suggestion
- ``hubble-export-file-max-size-mb``: size in MB at which to rotate the Hubble export file. (default 10)

- ``hubble-export-file-max-backups``: number of rotated Hubble export files to keep. (default 5)

- ``hubble-export-file-compress``: compress rotated Hubble export files.
```

Performance tuning
==================

> Configuration options impacting performance of **Hubble exporter**:

Reviewer 2: Clarity
```suggestion
Configuration options impacting performance of **Hubble exporter** include:
```

- ``hubble-export-allowlist``:
  Specify allowlist as JSON encoded FlowFilters to Hubble exporter.

- ``hubble-export-denylist``:
  Specify denylist as JSON encoded FlowFilters to Hubble exporter.

- ``hubble-export-fieldmask``:

>   Specify list of fields to use for field mask in Hubble exporter.

Reviewer 2: Again, these pairs work better as single lines.
```suggestion
- ``hubble-export-allowlist``: specify an allowlist as JSON encoded FlowFilters to Hubble exporter.

- ``hubble-export-denylist``: specify a denylist as JSON encoded FlowFilters to Hubble exporter.

- ``hubble-export-fieldmask``: specify a list of fields to use for field masking in Hubble exporter.
```

Filters
-------

> You can use ``hubble`` CLI to generated required filters (see `Specifying Raw

Reviewer 2: Good use of active voice and referring to the user as "you" 👍🏻

Flow Filters`_ for more examples).

.. _Specifying Raw Flow Filters: https://github.com/cilium/hubble#specifying-raw-flow-filters

> For example to filter flows with verdict ``DENIED`` or ``ERROR`` run:

Reviewer 2: Comma nits
```suggestion
For example, to filter flows with verdict ``DENIED`` or ``ERROR``, run:
```

.. code-block:: shell-session

    $ hubble observe --verdict DROPPED --verdict ERROR --print-raw-filters
    allowlist:
    - '{"verdict":["DROPPED","ERROR"]}'

> and paste the output to ``hubble-export-allowlist`` in ``cilium-config``

Reviewer 2: Grammar nit
```suggestion
Then paste the output to ``hubble-export-allowlist`` in ``cilium-config``
```

Config Map:

.. code-block:: shell-session

    kubectl -n kube-system patch cm cilium-config --patch-file=/dev/stdin <<-EOF
    data:
      hubble-export-allowlist: '{"verdict":["DROPPED","ERROR"]}'
    EOF

The same can be done to filter out some data. For example all flows in

> ``kube-system`` namespace:

Reviewer 2: Clarity, active voice, consistency with previous example
```suggestion
You can do the same to selectively filter data. For example, to filter all flows in the
``kube-system`` namespace, run:
```

.. code-block:: shell-session

    $ hubble observe --not --namespace kube-system --print-raw-filters
    denylist:
    - '{"source_pod":["kube-system/"]}'
    - '{"destination_pod":["kube-system/"]}'

> and paste the output to ``hubble-export-denylist`` in ``cilium-config`` Config

Reviewer 2: Grammar nit
```suggestion
Then paste the output to ``hubble-export-denylist`` in ``cilium-config`` Config
```

Map:

.. code-block:: shell-session

    kubectl -n kube-system patch cm cilium-config --patch-file=/dev/stdin <<-EOF
    data:
      hubble-export-denylist: '{"source_pod":["kube-system/"]},{"destination_pod":["kube-system/"]}'
    EOF

Field mask
----------

> Currently, field masks can't be generated with ``hubble``. It is a list of field names from `flow proto`_ definition.

Reviewer 2: Avoid references to the future or present as an ambiguous state. Specific is better than general: what does "it" refer to?
```suggestion
Field masks can't be generated with ``hubble``. It is a list of field names from the `flow proto`_ definition.
```

.. _flow proto: https://github.com/cilium/cilium/blob/main/api/v1/flow/flow.proto

> Examples:

Reviewer 2: Grammar
```suggestion
Examples include:
```

>  - Keep most of the information with notable exception of pod labels:

Reviewer 2: Grammar and clarity
```suggestion
 - To keep all information except pod labels:
```

   .. code-block:: shell-session

       hubble-export-fieldmask: time source.identity source.namespace source.pod_name destination.identity destination.namespace destination.pod_name source_service destination_service l4 IP ethernet l7 Type node_name is_reply event_type verdict Summary

>  - Keep only timestamp, verdict, ports, IP addresses, node name, pod name and namespace:

Reviewer 2: Grammar
```suggestion
 - To keep only timestamp, verdict, ports, IP addresses, node name, pod name, and namespace:
```

   .. code-block:: shell-session

       hubble-export-fieldmask: time source.namespace source.pod_name destination.namespace destination.pod_name l4 IP node_name is_reply verdict

> Complete example:

Reviewer 2: ```suggestion
The following is a complete example of configuring Hubble Exporter.
```
